<!DOCTYPE html><html lang="ko"><head><meta charSet="utf-8"/><title>Swift에서의 오류 처리 | ui-station</title><meta name="description" content=""/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><meta property="og:url" content="https://ui-station.github.io///post/2024-06-19-ErrorhandlinginSwift" data-gatsby-head="true"/><meta property="og:type" content="website" data-gatsby-head="true"/><meta property="og:site_name" content="Swift에서의 오류 처리 | ui-station" data-gatsby-head="true"/><meta property="og:title" content="Swift에서의 오류 처리 | ui-station" data-gatsby-head="true"/><meta property="og:description" content="" data-gatsby-head="true"/><meta property="og:image" content="/assets/img/2024-06-19-ErrorhandlinginSwift_0.png" data-gatsby-head="true"/><meta property="og:locale" content="en_US" data-gatsby-head="true"/><meta name="twitter:card" content="summary_large_image" data-gatsby-head="true"/><meta property="twitter:domain" content="https://ui-station.github.io/" data-gatsby-head="true"/><meta property="twitter:url" content="https://ui-station.github.io///post/2024-06-19-ErrorhandlinginSwift" data-gatsby-head="true"/><meta name="twitter:title" content="Swift에서의 오류 처리 | ui-station" data-gatsby-head="true"/><meta name="twitter:description" content="" data-gatsby-head="true"/><meta name="twitter:image" content="/assets/img/2024-06-19-ErrorhandlinginSwift_0.png" data-gatsby-head="true"/><meta name="twitter:data1" content="Dev | ui-station" data-gatsby-head="true"/><meta name="article:published_time" content="2024-06-19 11:08" data-gatsby-head="true"/><meta name="next-head-count" content="19"/><meta name="google-site-verification" content="a-yehRo3k3xv7fg6LqRaE8jlE42e5wP2bDE_2F849O4"/><link rel="stylesheet" href="/favicons/favicon.ico"/><link rel="icon" type="image/png" sizes="16x16" href="/assets/favicons/favicon-16x16.png"/><link rel="icon" type="image/png" sizes="32x32" href="/assets/favicons/favicon-32x32.png"/><link rel="icon" type="image/png" sizes="96x96" href="/assets/favicons/favicon-96x96.png"/><link rel="icon" href="/favicons/apple-icon-180x180.png"/><link rel="apple-touch-icon" href="/favicons/apple-icon-180x180.png"/><link rel="apple-touch-startup-image" href="/startup.png"/><meta name="apple-mobile-web-app-capable" content="yes"/><meta name="apple-mobile-web-app-status-bar-style" content="black"/><meta name="msapplication-config" content="/favicons/browserconfig.xml"/><script async="" src="https://www.googletagmanager.com/gtag/js?id=G-BHFR6GTH9P"></script><script>window.dataLayer = window.dataLayer || [];
            function gtag(){dataLayer.push(arguments);}
            gtag('js', new Date());
          
            gtag('config', 'G-BHFR6GTH9P');</script><link rel="preload" href="/_next/static/css/6e57edcf9f2ce551.css" as="style"/><link rel="stylesheet" href="/_next/static/css/6e57edcf9f2ce551.css" data-n-g=""/><link rel="preload" href="/_next/static/css/b8ef307c9aee1e34.css" as="style"/><link rel="stylesheet" href="/_next/static/css/b8ef307c9aee1e34.css" data-n-p=""/><noscript data-n-css=""></noscript><script defer="" nomodule="" src="/_next/static/chunks/polyfills-c67a75d1b6f99dc8.js"></script><script src="/_next/static/chunks/webpack-ee6df16fdc6dae4d.js" defer=""></script><script src="/_next/static/chunks/framework-46611630e39cfdeb.js" defer=""></script><script src="/_next/static/chunks/main-cf4a52eec9a970a0.js" defer=""></script><script src="/_next/static/chunks/pages/_app-6fae11262ee5c69b.js" defer=""></script><script src="/_next/static/chunks/75fc9c18-4a646156c659a948.js" defer=""></script><script src="/_next/static/chunks/348-d11c34b645b13f5b.js" defer=""></script><script src="/_next/static/chunks/162-4172e84c8e2aa747.js" defer=""></script><script src="/_next/static/chunks/pages/post/%5Bslug%5D-4f7b40c1114f0d09.js" defer=""></script><script src="/_next/static/YUMR4jSyk_WlOHHc7UfOk/_buildManifest.js" defer=""></script><script src="/_next/static/YUMR4jSyk_WlOHHc7UfOk/_ssgManifest.js" defer=""></script></head><body><div id="__next"><header class="Header_header__Z8PUO"><div class="Header_inner__tfr0u"><strong class="Header_title__Otn70"><a href="/">UI STATION</a></strong><nav class="Header_nav_area__6KVpk"><a class="nav_item" href="/posts/1">Posts</a></nav></div></header><main class="posts_container__NyRU3"><div class="posts_inner__i3n_i"><h1 class="posts_post_title__EbxNx">Swift에서의 오류 처리</h1><div class="posts_meta__cR7lu"><div class="posts_profile_wrap__mslMl"><div class="posts_profile_image_wrap__kPikV"><img alt="Swift에서의 오류 처리" loading="lazy" width="44" height="44" decoding="async" data-nimg="1" class="profile" style="color:transparent" src="/favicons/apple-icon-114x114.png"/></div><div class="posts_textarea__w_iKT"><span class="writer">UI STATION</span><span class="posts_info__5KJdN"><span class="posts_date__ctqHI">Posted On Jun 19, 2024</span><span class="posts_reading_time__f7YPP">10<!-- --> min read</span></span></div></div><img alt="" loading="lazy" width="50" height="50" decoding="async" data-nimg="1" class="posts_view_badge__tcbfm" style="color:transparent" src="https://hits.seeyoufarm.com/api/count/incr/badge.svg?url=https%3A%2F%2Fallround-coder.github.io/post/2024-06-19-ErrorhandlinginSwift&amp;count_bg=%2379C83D&amp;title_bg=%23555555&amp;icon=&amp;icon_color=%23E7E7E7&amp;title=views&amp;edge_flat=false"/></div><article class="posts_post_content__n_L6j"><div><!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta content="width=device-width, initial-scale=1" name="viewport">
</head>
<body>
<h2>사용자에게 경고하기 위해 do-catch 문과 사용자 정의 오류를 던지는 방법 (코드 예제와 함께)</h2>
<p>에러 처리는 모든 앱의 기본적인 부분입니다.</p>
<p>사용자가 수행하는 모든 조치가 성공하는 것은 아닙니다. 일부는 실패할 것이며, 우리의 앱은 사용자에게 일어난 일을 쉽게 알려주고 문제를 피하기 위해 무엇을 시도할 수 있는지 알려줘야 합니다.</p>
<p>간단한 throw 및 do-catch 문을 사용하여 간단한 에러 처리를 어떻게 달성할 수 있는지 보여드릴게요.</p>
<!-- ui-station 사각형 -->
<p><ins class="adsbygoogle" style="display:block" data-ad-client="ca-pub-4877378276818686" data-ad-slot="7249294152" data-ad-format="auto" data-full-width-responsive="true"></ins></p>
<script>
(adsbygoogle = window.adsbygoogle || []).push({});
</script>
<img src="/assets/img/2024-06-19-에러핸들링인Swif_0.png">
<h2>에러 정의 및 던지기</h2>
<p>에러를 표현하려면 Error 프로토콜을 준수하는 타입을 사용해야 합니다. 이는 아무런 제약이 없는 빈 프로토콜이며, 우리가 가장 잘 적합한 타입으로 에러를 정의할 수 있다는 것을 의미합니다.</p>
<p>일반적으로 enum이 시작하기에 가장 좋은 방법입니다. 네트워크 에러를 처리하고 싶다고 가정해보면, 아래 enum을 정의할 수 있습니다👇</p>
<!-- ui-station 사각형 -->
<p><ins class="adsbygoogle" style="display:block" data-ad-client="ca-pub-4877378276818686" data-ad-slot="7249294152" data-ad-format="auto" data-full-width-responsive="true"></ins></p>
<script>
(adsbygoogle = window.adsbygoogle || []).push({});
</script>
<pre><code class="hljs language-js">열거형 <span class="hljs-title class_">NetworkError</span>: 오류 {
    <span class="hljs-keyword">case</span> unexpected
    <span class="hljs-keyword">case</span> <span class="hljs-title function_">invalidURL</span>(_ <span class="hljs-attr">url</span>: <span class="hljs-variable constant_">URL</span>)
    <span class="hljs-keyword">case</span> <span class="hljs-title function_">apiError</span>(<span class="hljs-attr">statusCode</span>: <span class="hljs-title class_">Int</span>)
}
</code></pre>
<p>백엔드 API로 요청을 수행할 때 오류가 발생하면 NetworkError를 던질 수 있습니다.</p>
<pre><code class="hljs language-js">열거형 <span class="hljs-title class_">ApiMethod</span> {...}
구조체 <span class="hljs-title class_">ApiRequest</span> {...}

func <span class="hljs-title function_">createRequest</span>(<span class="hljs-keyword">from</span> <span class="hljs-attr">url</span>: <span class="hljs-title class_">String</span>, <span class="hljs-attr">method</span>: <span class="hljs-title class_">ApiMethod</span>) throws -> <span class="hljs-title class_">ApiRequest</span> {
  guard <span class="hljs-keyword">let</span> <span class="hljs-variable constant_">URL</span> = <span class="hljs-title function_">URL</span>(<span class="hljs-attr">string</span>: url) <span class="hljs-keyword">else</span> {
    <span class="hljs-keyword">throw</span> <span class="hljs-title class_">NetworkError</span>.<span class="hljs-title function_">invalidURL</span>(url)
  }
  ...
}
</code></pre>
<h2>오류 처리</h2>
<!-- ui-station 사각형 -->
<p><ins class="adsbygoogle" style="display:block" data-ad-client="ca-pub-4877378276818686" data-ad-slot="7249294152" data-ad-format="auto" data-full-width-responsive="true"></ins></p>
<script>
(adsbygoogle = window.adsbygoogle || []).push({});
</script>
<p>우리는 사용자 정의 오류를 갖고 있고, 이를 throw하고 있습니다. 하지만, 이를 어떻게 잡을 수 있을까요?</p>
<p>다른 많은 언어와 마찬가지로, Swift에는 do-catch 문이 있습니다. 만약 do 절 내에서 오류가 throw되면, 실행은 catch 절로 전달되고 오류는 로컬 상수 error에 저장됩니다.</p>
<pre><code class="hljs language-js"><span class="hljs-keyword">do</span> {
  <span class="hljs-keyword">let</span> result = <span class="hljs-keyword">try</span> <span class="hljs-title function_">createRequest</span>(<span class="hljs-attr">from</span>: <span class="hljs-string">"//myapi.com/"</span>, <span class="hljs-attr">method</span>: .<span class="hljs-property">GET</span>)
} <span class="hljs-keyword">catch</span> {
  <span class="hljs-title function_">print</span>(error)
  <span class="hljs-comment">// 적절한 처리를 수행</span>
}
</code></pre>
<p>가끔씩 우리는 잡은 오류 형식(또는 오류 값)에 따라 특정 작업을 트리거하길 원합니다. 이를 패턴을 사용하여 수행할 수 있습니다.</p>
<!-- ui-station 사각형 -->
<p><ins class="adsbygoogle" style="display:block" data-ad-client="ca-pub-4877378276818686" data-ad-slot="7249294152" data-ad-format="auto" data-full-width-responsive="true"></ins></p>
<script>
(adsbygoogle = window.adsbygoogle || []).push({});
</script>
<pre><code class="hljs language-js"><span class="hljs-keyword">do</span> {
  <span class="hljs-keyword">let</span> result = <span class="hljs-keyword">try</span> <span class="hljs-title function_">makeRequest</span>(<span class="hljs-attr">to</span>: <span class="hljs-string">"//myapi.com/"</span>, <span class="hljs-attr">method</span>: .<span class="hljs-property">GET</span>)
} <span class="hljs-keyword">catch</span> <span class="hljs-title class_">NetworkError</span>.<span class="hljs-title function_">invalidUrl</span>(<span class="hljs-params"><span class="hljs-keyword">let</span> url</span>) {
  <span class="hljs-comment">// ...</span>
} <span class="hljs-keyword">catch</span> {
  <span class="hljs-title function_">print</span>(error)
  <span class="hljs-comment">//</span>
}
</code></pre>
<h2>에러 전파하기</h2>
<p>던져진 함수는 함수 내에서 발생한 에러를 호출자의 범위로 전파할 수 있습니다.</p>
<pre><code class="hljs language-js">struct <span class="hljs-title class_">User</span>: <span class="hljs-title class_">Decodable</span> {
    <span class="hljs-keyword">let</span> <span class="hljs-attr">name</span>: <span class="hljs-title class_">String</span>
    <span class="hljs-keyword">let</span> <span class="hljs-attr">age</span>: <span class="hljs-title class_">Int</span>
}

func decodeJSON&#x3C;<span class="hljs-attr">T</span>: <span class="hljs-title class_">Decodable</span>>(_ <span class="hljs-attr">jsonString</span>: <span class="hljs-title class_">String</span>) throws -> T {
    <span class="hljs-keyword">let</span> jsonData = <span class="hljs-title class_">Data</span>(jsonString.<span class="hljs-property">utf8</span>)
    <span class="hljs-keyword">let</span> decodedObject = <span class="hljs-keyword">try</span> <span class="hljs-title class_">JSON</span>Decoder().<span class="hljs-title function_">decode</span>(T.<span class="hljs-property">self</span>, <span class="hljs-attr">from</span>: jsonData)
    <span class="hljs-keyword">return</span> decodedObject
}

func <span class="hljs-title function_">getUserFromJSON</span>(_ <span class="hljs-attr">jsonString</span>: <span class="hljs-title class_">String</span>) -> <span class="hljs-title class_">User</span>? {
    <span class="hljs-keyword">do</span> {
        <span class="hljs-keyword">let</span> <span class="hljs-attr">user</span>: <span class="hljs-title class_">User</span> = <span class="hljs-keyword">try</span> <span class="hljs-title function_">decodeJSON</span>(jsonString)
        <span class="hljs-keyword">return</span> user
    } <span class="hljs-keyword">catch</span> {
        <span class="hljs-title function_">print</span>(error)
        <span class="hljs-keyword">return</span> nil
    }
}
</code></pre>
<!-- ui-station 사각형 -->
<p><ins class="adsbygoogle" style="display:block" data-ad-client="ca-pub-4877378276818686" data-ad-slot="7249294152" data-ad-format="auto" data-full-width-responsive="true"></ins></p>
<script>
(adsbygoogle = window.adsbygoogle || []).push({});
</script>
<h2>try? &#x26; try!</h2>
<p>가끔은 던지는 함수로 발생한 오류에 대해 신경 쓰지 않을 수 있습니다. 이런 상황에서는 do-catch 문을 사용하지 않고 try? 지시어를 사용할 수 있습니다.</p>
<p>try?를 사용하면 호출하는 함수가 옵셔널이 됩니다. 즉, 반환 값을 가진 함수를 호출하는 경우 값은 nil이 됩니다.</p>
<pre><code class="hljs language-js"><span class="hljs-keyword">let</span> request = <span class="hljs-keyword">try</span>? <span class="hljs-title function_">createRequest</span>(<span class="hljs-attr">from</span>: <span class="hljs-string">"malformedurl"</span>, <span class="hljs-attr">method</span>: .<span class="hljs-property">GET</span>)
<span class="hljs-comment">// request = nil</span>
</code></pre>
<!-- ui-station 사각형 -->
<p><ins class="adsbygoogle" style="display:block" data-ad-client="ca-pub-4877378276818686" data-ad-slot="7249294152" data-ad-format="auto" data-full-width-responsive="true"></ins></p>
<script>
(adsbygoogle = window.adsbygoogle || []).push({});
</script>
<p>한편, 때때로 특정한 던지기 함수가 런타임에서 실패하지 않을 것을 확신할 수 있습니다. 이런 경우에는 try!를 사용할 수 있습니다. 그리고 try?와 마찬가지로 do-catch 클로저를 사용할 필요가 없습니다.</p>
<pre><code class="hljs language-js"><span class="hljs-keyword">let</span> request = <span class="hljs-keyword">try</span>! <span class="hljs-title function_">createRequest</span>(<span class="hljs-attr">from</span>: <span class="hljs-string">"https://myapi.com/"</span>, <span class="hljs-attr">method</span>: .<span class="hljs-property">GET</span>)
</code></pre>
<h2>비동기 코드</h2>
<p>에러를 던지는 것은 비동기 코드와 잘 작동합니다. 함수 선언에서 throws 키워드 앞에 async 키워드를 넣어주어야 합니다.</p>
<!-- ui-station 사각형 -->
<p><ins class="adsbygoogle" style="display:block" data-ad-client="ca-pub-4877378276818686" data-ad-slot="7249294152" data-ad-format="auto" data-full-width-responsive="true"></ins></p>
<script>
(adsbygoogle = window.adsbygoogle || []).push({});
</script>
<p>비동기 함수를 호출할 때는 try await (키워드 순서 스와이프)를 사용합니다.</p>
<pre><code class="hljs language-js">func <span class="hljs-title function_">performRequest</span>(<span class="hljs-keyword">with</span> <span class="hljs-attr">request</span>: <span class="hljs-title class_">ApiRequest</span>) <span class="hljs-keyword">async</span> throws -> <span class="hljs-title class_">ApiResult</span> {...}

<span class="hljs-comment">// ----------------</span>

<span class="hljs-keyword">do</span> {
  <span class="hljs-keyword">let</span> request = <span class="hljs-keyword">try</span> <span class="hljs-title function_">makeRequest</span>(<span class="hljs-attr">to</span>: <span class="hljs-string">"https://myapi.com/users"</span>, <span class="hljs-attr">method</span>: .<span class="hljs-property">GET</span>)
  <span class="hljs-keyword">let</span> apiRestul = <span class="hljs-keyword">try</span> <span class="hljs-keyword">await</span> <span class="hljs-title function_">performRequest</span>(<span class="hljs-attr">with</span>: request)
} <span class="hljs-keyword">catch</span> <span class="hljs-title class_">NetworkError</span>.<span class="hljs-title function_">invalidUrl</span>(<span class="hljs-params"><span class="hljs-keyword">let</span> url</span>) {
  <span class="hljs-comment">// ...</span>
} <span class="hljs-keyword">catch</span> {
  <span class="hljs-title function_">print</span>(error)
  <span class="hljs-comment">//</span>
}
</code></pre>
<h1>사용자에게 알림</h1>
<p>에러 구조를 구축했으니, 이제 사용자들에게 무엇이 잘못되었는지 알려줄 필요가 있습니다.</p>
<!-- ui-station 사각형 -->
<p><ins class="adsbygoogle" style="display:block" data-ad-client="ca-pub-4877378276818686" data-ad-slot="7249294152" data-ad-format="auto" data-full-width-responsive="true"></ins></p>
<script>
(adsbygoogle = window.adsbygoogle || []).push({});
</script>
<img src="https://miro.medium.com/v2/resize:fit:592/1*cfrDQghry5EPpPhMSFFDaA.gif">
<p>각각의 오류에 대한 설명 은 오류가 처음에 발생한 이유입니다.
사용자가 사용자를 피할 수 있는 가능한 조치입니다.</p>
<p>우리는 Error enum에 해당 정보를 변수로 추가할 수 있습니다. 그러나 이미 모든 이 정보를 그룹화하기 위한 프로토콜이 있습니다: LocalizedError. 우리의 사용 사례에서 우리는 failureReason 및 recoverySuggestion 변수 만 구현하면 됩니다.</p>
<!-- ui-station 사각형 -->
<p><ins class="adsbygoogle" style="display:block" data-ad-client="ca-pub-4877378276818686" data-ad-slot="7249294152" data-ad-format="auto" data-full-width-responsive="true"></ins></p>
<script>
(adsbygoogle = window.adsbygoogle || []).push({});
</script>
<pre><code class="hljs language-swift"><span class="hljs-keyword">enum</span> <span class="hljs-title class_">OrderError</span>: <span class="hljs-title class_">LocalizedError</span> {
  <span class="hljs-keyword">case</span> unexpected
  <span class="hljs-keyword">case</span> outOfCoffee(coffee: <span class="hljs-type">Coffee</span>)
  <span class="hljs-keyword">case</span> minimumNotMet(currentPrice: <span class="hljs-type">Float</span>)

  <span class="hljs-keyword">var</span> failureReason: <span class="hljs-type">String</span>? {
    <span class="hljs-keyword">switch</span> <span class="hljs-keyword">self</span> {
    <span class="hljs-keyword">case</span> .unexpected:
      <span class="hljs-keyword">return</span> <span class="hljs-string">"주문 처리 중 문제가 발생했습니다."</span>
    <span class="hljs-keyword">case</span> .outOfCoffee(<span class="hljs-keyword">let</span> coffee):
      <span class="hljs-keyword">return</span> <span class="hljs-string">"<span class="hljs-subst">\(coffee.name)</span>가 다 떨어졌어요."</span>
    <span class="hljs-keyword">case</span> .minimumNotMet:
      <span class="hljs-keyword">return</span> <span class="hljs-string">"앱에서 주문하려면 최소 $10 이상 사용해야 합니다."</span>
    }
  }

  <span class="hljs-keyword">var</span> recoverySuggestion: <span class="hljs-type">String</span>? {
    <span class="hljs-keyword">switch</span> <span class="hljs-keyword">self</span> {
    <span class="hljs-keyword">case</span> .unexpected:
      <span class="hljs-keyword">return</span> <span class="hljs-string">"몇 분 후에 다시 시도해주세요."</span>
    <span class="hljs-keyword">case</span> .outOfCoffee:
      <span class="hljs-keyword">return</span> <span class="hljs-string">"다른 종류의 커피로 변경해보세요."</span>
    <span class="hljs-keyword">case</span> .minimumNotMet(<span class="hljs-keyword">let</span> currentPrice):
      <span class="hljs-keyword">return</span> <span class="hljs-string">"주문에 $<span class="hljs-subst">\(<span class="hljs-number">10</span><span class="hljs-operator">-</span>currentPrice)</span> 더 추가해주세요."</span>
    }
  }
}
</code></pre>
<p>여기에서 작업이 완료됩니다. 그러나 앱에 통합된 로깅 솔루션이 있다면, 오류 구조를 확장하여 추가 정보를 추가할 수 있습니다.</p>
<pre><code class="hljs language-swift"><span class="hljs-keyword">import</span> Foundation
<span class="hljs-keyword">import</span> HKLogger

<span class="hljs-keyword">struct</span> <span class="hljs-title class_">AppError</span>: <span class="hljs-title class_">Error</span> {
  <span class="hljs-keyword">let</span> type: <span class="hljs-type">LocalizedError</span>
  <span class="hljs-keyword">var</span> userMessage: <span class="hljs-type">String</span> {
    <span class="hljs-keyword">return</span> <span class="hljs-string">"<span class="hljs-subst">\(type.failureReason <span class="hljs-operator">??</span> ErrorConstants.defaultError)</span> <span class="hljs-subst">\n</span><span class="hljs-subst">\n</span> <span class="hljs-subst">\(type.recoverySuggestion <span class="hljs-operator">??</span> ErrorConstants.defaultAction)</span>"</span>
  }
  <span class="hljs-comment">// 필요한 만큼 많은 속성을 추가</span>

  <span class="hljs-keyword">init</span>(<span class="hljs-params">type</span>: <span class="hljs-type">LocalizedError</span>, <span class="hljs-params">debugInfo</span>: <span class="hljs-type">String</span>? <span class="hljs-operator">=</span> <span class="hljs-literal">nil</span>) {
      <span class="hljs-keyword">self</span>.type <span class="hljs-operator">=</span> type
      <span class="hljs-keyword">guard</span> <span class="hljs-keyword">let</span> debugInfo <span class="hljs-keyword">else</span> { <span class="hljs-keyword">return</span> }
      <span class="hljs-type">HKLogger</span>.shared.error(message: debugInfo)
  }
}
</code></pre>
<p>이렇게 함으로써 로깅 오류 로직을 한 곳에 중앙 집중화할 수 있습니다.</p>
<!-- ui-station 사각형 -->
<p><ins class="adsbygoogle" style="display:block" data-ad-client="ca-pub-4877378276818686" data-ad-slot="7249294152" data-ad-format="auto" data-full-width-responsive="true"></ins></p>
<script>
(adsbygoogle = window.adsbygoogle || []).push({});
</script>
<p>마침내 사용자 정의 AppError를 사용하여 경고를 표시하는 간단한 View 확장을 만들 수 있습니다.</p>
<pre><code class="hljs language-js">extension <span class="hljs-title class_">View</span> {
    func <span class="hljs-title function_">alert</span>(<span class="hljs-attr">isPresented</span>: <span class="hljs-title class_">Binding</span>&#x3C;<span class="hljs-title class_">Bool</span>>, withError <span class="hljs-attr">error</span>: <span class="hljs-title class_">AppError</span>?) -> some <span class="hljs-title class_">View</span> {
        <span class="hljs-keyword">return</span> <span class="hljs-title function_">alert</span>(
            <span class="hljs-string">"Ups! :("</span>,
            <span class="hljs-attr">isPresented</span>: isPresented,
            <span class="hljs-attr">actions</span>: {
                <span class="hljs-title class_">Button</span>(<span class="hljs-string">"Ok"</span>) {}
            }, <span class="hljs-attr">message</span>: {
                <span class="hljs-title class_">Text</span>(error?.<span class="hljs-property">userMessage</span> ?? <span class="hljs-string">""</span>)
            }
        )
    }
}

<span class="hljs-comment">// ---------------------------------------------------</span>

struct <span class="hljs-title class_">OrderConfirmationView</span>: <span class="hljs-title class_">View</span> {
  @<span class="hljs-title class_">StateObject</span> private <span class="hljs-keyword">var</span> viewModel = <span class="hljs-title class_">ConfirmationViewModel</span>()

  <span class="hljs-keyword">var</span> <span class="hljs-attr">body</span>: some <span class="hljs-title class_">View</span> {
    <span class="hljs-title class_">VStack</span>(<span class="hljs-attr">alignment</span>: .<span class="hljs-property">leading</span>) {
      <span class="hljs-comment">// ...</span>
      <span class="hljs-title class_">Button</span>(<span class="hljs-string">"주문하기"</span>) {
        viewModel.<span class="hljs-title function_">tryToPlaceOrder</span>(order)
      }
    }
    .<span class="hljs-title function_">alert</span>(<span class="hljs-attr">isPresented</span>: $viewModel.<span class="hljs-property">showError</span>, <span class="hljs-attr">withError</span>: viewModel.<span class="hljs-property">appError</span>)
  }
}

<span class="hljs-comment">// ---------------------------------------------------</span>

@<span class="hljs-title class_">Observable</span>
final <span class="hljs-keyword">class</span> <span class="hljs-title class_">ConfirmationViewModel</span>: <span class="hljs-title class_">ObservableObject</span> {
    <span class="hljs-keyword">var</span> <span class="hljs-attr">appError</span>: <span class="hljs-title class_">AppError</span>?
    <span class="hljs-keyword">var</span> showError = <span class="hljs-literal">false</span>

    func <span class="hljs-title function_">tryToPlaceOrder</span>(<span class="hljs-params">_ order: Order</span>) {
        <span class="hljs-keyword">do</span> {
            <span class="hljs-keyword">try</span> <span class="hljs-title class_">OrdersManager</span>.<span class="hljs-property">shared</span>.<span class="hljs-title function_">add</span>(order)
        } <span class="hljs-keyword">catch</span> <span class="hljs-keyword">let</span> error <span class="hljs-keyword">as</span> <span class="hljs-title class_">AppError</span> {
            showError = <span class="hljs-literal">true</span>
            appError = error
        } <span class="hljs-keyword">catch</span> {
            appError = <span class="hljs-title class_">AppError</span>(<span class="hljs-attr">type</span>: <span class="hljs-title class_">OrderError</span>.<span class="hljs-property">unexpected</span>)
        }
    }
}

<span class="hljs-comment">// ---------------------------------------------------</span>

final <span class="hljs-keyword">class</span> <span class="hljs-title class_">OrdersManager</span> {
  <span class="hljs-comment">// ...</span>
  func <span class="hljs-title function_">add</span>(_ <span class="hljs-attr">order</span>: <span class="hljs-title class_">Order</span>) throws {
    <span class="hljs-keyword">if</span> order.<span class="hljs-property">price</span> >= <span class="hljs-number">10</span> {
      orders.<span class="hljs-title function_">append</span>(order)
    } <span class="hljs-keyword">else</span> {
      <span class="hljs-keyword">throw</span> <span class="hljs-title class_">AppError</span>(<span class="hljs-attr">type</span>: <span class="hljs-title class_">OrderError</span>.<span class="hljs-title function_">minimumNotMet</span>(<span class="hljs-attr">currentPrice</span>: order.<span class="hljs-property">price</span>))
    }
  }
}
</code></pre>
<p>질문이 있으신가요? 언제든지 메시지를 남겨주세요! 🙂</p>
<ul>
<li>🤓 iOS 개발 팁과 통찰력에 관한 콘텐츠를 X에서 함께 확인해보세요.</li>
<li>🚀 모든 예제 프로젝트를 공유하는 내 GitHub를 확인해보세요 (이것도 포함됩니다 😉)</li>
</ul>
</body>
</html>
</div></article></div></main></div><script id="__NEXT_DATA__" type="application/json">{"props":{"pageProps":{"post":{"title":"Swift에서의 오류 처리","description":"","date":"2024-06-19 11:08","slug":"2024-06-19-ErrorhandlinginSwift","content":"\n## 사용자에게 경고하기 위해 do-catch 문과 사용자 정의 오류를 던지는 방법 (코드 예제와 함께)\n\n에러 처리는 모든 앱의 기본적인 부분입니다.\n\n사용자가 수행하는 모든 조치가 성공하는 것은 아닙니다. 일부는 실패할 것이며, 우리의 앱은 사용자에게 일어난 일을 쉽게 알려주고 문제를 피하기 위해 무엇을 시도할 수 있는지 알려줘야 합니다.\n\n간단한 throw 및 do-catch 문을 사용하여 간단한 에러 처리를 어떻게 달성할 수 있는지 보여드릴게요.\n\n\u003c!-- ui-station 사각형 --\u003e\n\n\u003cins class=\"adsbygoogle\"\nstyle=\"display:block\"\ndata-ad-client=\"ca-pub-4877378276818686\"\ndata-ad-slot=\"7249294152\"\ndata-ad-format=\"auto\"\ndata-full-width-responsive=\"true\"\u003e\u003c/ins\u003e\n\n\u003cscript\u003e\n(adsbygoogle = window.adsbygoogle || []).push({});\n\u003c/script\u003e\n\n\u003cimg src=\"/assets/img/2024-06-19-에러핸들링인Swif_0.png\" /\u003e\n\n## 에러 정의 및 던지기\n\n에러를 표현하려면 Error 프로토콜을 준수하는 타입을 사용해야 합니다. 이는 아무런 제약이 없는 빈 프로토콜이며, 우리가 가장 잘 적합한 타입으로 에러를 정의할 수 있다는 것을 의미합니다.\n\n일반적으로 enum이 시작하기에 가장 좋은 방법입니다. 네트워크 에러를 처리하고 싶다고 가정해보면, 아래 enum을 정의할 수 있습니다👇\n\n\u003c!-- ui-station 사각형 --\u003e\n\n\u003cins class=\"adsbygoogle\"\nstyle=\"display:block\"\ndata-ad-client=\"ca-pub-4877378276818686\"\ndata-ad-slot=\"7249294152\"\ndata-ad-format=\"auto\"\ndata-full-width-responsive=\"true\"\u003e\u003c/ins\u003e\n\n\u003cscript\u003e\n(adsbygoogle = window.adsbygoogle || []).push({});\n\u003c/script\u003e\n\n```js\n열거형 NetworkError: 오류 {\n    case unexpected\n    case invalidURL(_ url: URL)\n    case apiError(statusCode: Int)\n}\n```\n\n백엔드 API로 요청을 수행할 때 오류가 발생하면 NetworkError를 던질 수 있습니다.\n\n```js\n열거형 ApiMethod {...}\n구조체 ApiRequest {...}\n\nfunc createRequest(from url: String, method: ApiMethod) throws -\u003e ApiRequest {\n  guard let URL = URL(string: url) else {\n    throw NetworkError.invalidURL(url)\n  }\n  ...\n}\n```\n\n## 오류 처리\n\n\u003c!-- ui-station 사각형 --\u003e\n\n\u003cins class=\"adsbygoogle\"\nstyle=\"display:block\"\ndata-ad-client=\"ca-pub-4877378276818686\"\ndata-ad-slot=\"7249294152\"\ndata-ad-format=\"auto\"\ndata-full-width-responsive=\"true\"\u003e\u003c/ins\u003e\n\n\u003cscript\u003e\n(adsbygoogle = window.adsbygoogle || []).push({});\n\u003c/script\u003e\n\n우리는 사용자 정의 오류를 갖고 있고, 이를 throw하고 있습니다. 하지만, 이를 어떻게 잡을 수 있을까요?\n\n다른 많은 언어와 마찬가지로, Swift에는 do-catch 문이 있습니다. 만약 do 절 내에서 오류가 throw되면, 실행은 catch 절로 전달되고 오류는 로컬 상수 error에 저장됩니다.\n\n```js\ndo {\n  let result = try createRequest(from: \"//myapi.com/\", method: .GET)\n} catch {\n  print(error)\n  // 적절한 처리를 수행\n}\n```\n\n가끔씩 우리는 잡은 오류 형식(또는 오류 값)에 따라 특정 작업을 트리거하길 원합니다. 이를 패턴을 사용하여 수행할 수 있습니다.\n\n\u003c!-- ui-station 사각형 --\u003e\n\n\u003cins class=\"adsbygoogle\"\nstyle=\"display:block\"\ndata-ad-client=\"ca-pub-4877378276818686\"\ndata-ad-slot=\"7249294152\"\ndata-ad-format=\"auto\"\ndata-full-width-responsive=\"true\"\u003e\u003c/ins\u003e\n\n\u003cscript\u003e\n(adsbygoogle = window.adsbygoogle || []).push({});\n\u003c/script\u003e\n\n```js\ndo {\n  let result = try makeRequest(to: \"//myapi.com/\", method: .GET)\n} catch NetworkError.invalidUrl(let url) {\n  // ...\n} catch {\n  print(error)\n  //\n}\n```\n\n## 에러 전파하기\n\n던져진 함수는 함수 내에서 발생한 에러를 호출자의 범위로 전파할 수 있습니다.\n\n```js\nstruct User: Decodable {\n    let name: String\n    let age: Int\n}\n\nfunc decodeJSON\u003cT: Decodable\u003e(_ jsonString: String) throws -\u003e T {\n    let jsonData = Data(jsonString.utf8)\n    let decodedObject = try JSONDecoder().decode(T.self, from: jsonData)\n    return decodedObject\n}\n\nfunc getUserFromJSON(_ jsonString: String) -\u003e User? {\n    do {\n        let user: User = try decodeJSON(jsonString)\n        return user\n    } catch {\n        print(error)\n        return nil\n    }\n}\n```\n\n\u003c!-- ui-station 사각형 --\u003e\n\n\u003cins class=\"adsbygoogle\"\nstyle=\"display:block\"\ndata-ad-client=\"ca-pub-4877378276818686\"\ndata-ad-slot=\"7249294152\"\ndata-ad-format=\"auto\"\ndata-full-width-responsive=\"true\"\u003e\u003c/ins\u003e\n\n\u003cscript\u003e\n(adsbygoogle = window.adsbygoogle || []).push({});\n\u003c/script\u003e\n\n## try? \u0026 try!\n\n가끔은 던지는 함수로 발생한 오류에 대해 신경 쓰지 않을 수 있습니다. 이런 상황에서는 do-catch 문을 사용하지 않고 try? 지시어를 사용할 수 있습니다.\n\ntry?를 사용하면 호출하는 함수가 옵셔널이 됩니다. 즉, 반환 값을 가진 함수를 호출하는 경우 값은 nil이 됩니다.\n\n```js\nlet request = try? createRequest(from: \"malformedurl\", method: .GET)\n// request = nil\n```\n\n\u003c!-- ui-station 사각형 --\u003e\n\n\u003cins class=\"adsbygoogle\"\nstyle=\"display:block\"\ndata-ad-client=\"ca-pub-4877378276818686\"\ndata-ad-slot=\"7249294152\"\ndata-ad-format=\"auto\"\ndata-full-width-responsive=\"true\"\u003e\u003c/ins\u003e\n\n\u003cscript\u003e\n(adsbygoogle = window.adsbygoogle || []).push({});\n\u003c/script\u003e\n\n한편, 때때로 특정한 던지기 함수가 런타임에서 실패하지 않을 것을 확신할 수 있습니다. 이런 경우에는 try!를 사용할 수 있습니다. 그리고 try?와 마찬가지로 do-catch 클로저를 사용할 필요가 없습니다.\n\n```js\nlet request = try! createRequest(from: \"https://myapi.com/\", method: .GET)\n```\n\n## 비동기 코드\n\n에러를 던지는 것은 비동기 코드와 잘 작동합니다. 함수 선언에서 throws 키워드 앞에 async 키워드를 넣어주어야 합니다.\n\n\u003c!-- ui-station 사각형 --\u003e\n\n\u003cins class=\"adsbygoogle\"\nstyle=\"display:block\"\ndata-ad-client=\"ca-pub-4877378276818686\"\ndata-ad-slot=\"7249294152\"\ndata-ad-format=\"auto\"\ndata-full-width-responsive=\"true\"\u003e\u003c/ins\u003e\n\n\u003cscript\u003e\n(adsbygoogle = window.adsbygoogle || []).push({});\n\u003c/script\u003e\n\n비동기 함수를 호출할 때는 try await (키워드 순서 스와이프)를 사용합니다.\n\n```js\nfunc performRequest(with request: ApiRequest) async throws -\u003e ApiResult {...}\n\n// ----------------\n\ndo {\n  let request = try makeRequest(to: \"https://myapi.com/users\", method: .GET)\n  let apiRestul = try await performRequest(with: request)\n} catch NetworkError.invalidUrl(let url) {\n  // ...\n} catch {\n  print(error)\n  //\n}\n```\n\n# 사용자에게 알림\n\n에러 구조를 구축했으니, 이제 사용자들에게 무엇이 잘못되었는지 알려줄 필요가 있습니다.\n\n\u003c!-- ui-station 사각형 --\u003e\n\n\u003cins class=\"adsbygoogle\"\nstyle=\"display:block\"\ndata-ad-client=\"ca-pub-4877378276818686\"\ndata-ad-slot=\"7249294152\"\ndata-ad-format=\"auto\"\ndata-full-width-responsive=\"true\"\u003e\u003c/ins\u003e\n\n\u003cscript\u003e\n(adsbygoogle = window.adsbygoogle || []).push({});\n\u003c/script\u003e\n\n\u003cimg src=\"https://miro.medium.com/v2/resize:fit:592/1*cfrDQghry5EPpPhMSFFDaA.gif\" /\u003e\n\n각각의 오류에 대한 설명 은 오류가 처음에 발생한 이유입니다.\n사용자가 사용자를 피할 수 있는 가능한 조치입니다.\n\n우리는 Error enum에 해당 정보를 변수로 추가할 수 있습니다. 그러나 이미 모든 이 정보를 그룹화하기 위한 프로토콜이 있습니다: LocalizedError. 우리의 사용 사례에서 우리는 failureReason 및 recoverySuggestion 변수 만 구현하면 됩니다.\n\n\u003c!-- ui-station 사각형 --\u003e\n\n\u003cins class=\"adsbygoogle\"\nstyle=\"display:block\"\ndata-ad-client=\"ca-pub-4877378276818686\"\ndata-ad-slot=\"7249294152\"\ndata-ad-format=\"auto\"\ndata-full-width-responsive=\"true\"\u003e\u003c/ins\u003e\n\n\u003cscript\u003e\n(adsbygoogle = window.adsbygoogle || []).push({});\n\u003c/script\u003e\n\n```swift\nenum OrderError: LocalizedError {\n  case unexpected\n  case outOfCoffee(coffee: Coffee)\n  case minimumNotMet(currentPrice: Float)\n\n  var failureReason: String? {\n    switch self {\n    case .unexpected:\n      return \"주문 처리 중 문제가 발생했습니다.\"\n    case .outOfCoffee(let coffee):\n      return \"\\(coffee.name)가 다 떨어졌어요.\"\n    case .minimumNotMet:\n      return \"앱에서 주문하려면 최소 $10 이상 사용해야 합니다.\"\n    }\n  }\n\n  var recoverySuggestion: String? {\n    switch self {\n    case .unexpected:\n      return \"몇 분 후에 다시 시도해주세요.\"\n    case .outOfCoffee:\n      return \"다른 종류의 커피로 변경해보세요.\"\n    case .minimumNotMet(let currentPrice):\n      return \"주문에 $\\(10-currentPrice) 더 추가해주세요.\"\n    }\n  }\n}\n```\n\n여기에서 작업이 완료됩니다. 그러나 앱에 통합된 로깅 솔루션이 있다면, 오류 구조를 확장하여 추가 정보를 추가할 수 있습니다.\n\n```swift\nimport Foundation\nimport HKLogger\n\nstruct AppError: Error {\n  let type: LocalizedError\n  var userMessage: String {\n    return \"\\(type.failureReason ?? ErrorConstants.defaultError) \\n\\n \\(type.recoverySuggestion ?? ErrorConstants.defaultAction)\"\n  }\n  // 필요한 만큼 많은 속성을 추가\n\n  init(type: LocalizedError, debugInfo: String? = nil) {\n      self.type = type\n      guard let debugInfo else { return }\n      HKLogger.shared.error(message: debugInfo)\n  }\n}\n```\n\n이렇게 함으로써 로깅 오류 로직을 한 곳에 중앙 집중화할 수 있습니다.\n\n\u003c!-- ui-station 사각형 --\u003e\n\n\u003cins class=\"adsbygoogle\"\nstyle=\"display:block\"\ndata-ad-client=\"ca-pub-4877378276818686\"\ndata-ad-slot=\"7249294152\"\ndata-ad-format=\"auto\"\ndata-full-width-responsive=\"true\"\u003e\u003c/ins\u003e\n\n\u003cscript\u003e\n(adsbygoogle = window.adsbygoogle || []).push({});\n\u003c/script\u003e\n\n마침내 사용자 정의 AppError를 사용하여 경고를 표시하는 간단한 View 확장을 만들 수 있습니다.\n\n```js\nextension View {\n    func alert(isPresented: Binding\u003cBool\u003e, withError error: AppError?) -\u003e some View {\n        return alert(\n            \"Ups! :(\",\n            isPresented: isPresented,\n            actions: {\n                Button(\"Ok\") {}\n            }, message: {\n                Text(error?.userMessage ?? \"\")\n            }\n        )\n    }\n}\n\n// ---------------------------------------------------\n\nstruct OrderConfirmationView: View {\n  @StateObject private var viewModel = ConfirmationViewModel()\n\n  var body: some View {\n    VStack(alignment: .leading) {\n      // ...\n      Button(\"주문하기\") {\n        viewModel.tryToPlaceOrder(order)\n      }\n    }\n    .alert(isPresented: $viewModel.showError, withError: viewModel.appError)\n  }\n}\n\n// ---------------------------------------------------\n\n@Observable\nfinal class ConfirmationViewModel: ObservableObject {\n    var appError: AppError?\n    var showError = false\n\n    func tryToPlaceOrder(_ order: Order) {\n        do {\n            try OrdersManager.shared.add(order)\n        } catch let error as AppError {\n            showError = true\n            appError = error\n        } catch {\n            appError = AppError(type: OrderError.unexpected)\n        }\n    }\n}\n\n// ---------------------------------------------------\n\nfinal class OrdersManager {\n  // ...\n  func add(_ order: Order) throws {\n    if order.price \u003e= 10 {\n      orders.append(order)\n    } else {\n      throw AppError(type: OrderError.minimumNotMet(currentPrice: order.price))\n    }\n  }\n}\n```\n\n질문이 있으신가요? 언제든지 메시지를 남겨주세요! 🙂\n\n- 🤓 iOS 개발 팁과 통찰력에 관한 콘텐츠를 X에서 함께 확인해보세요.\n- 🚀 모든 예제 프로젝트를 공유하는 내 GitHub를 확인해보세요 (이것도 포함됩니다 😉)\n","ogImage":{"url":"/assets/img/2024-06-19-ErrorhandlinginSwift_0.png"},"coverImage":"/assets/img/2024-06-19-ErrorhandlinginSwift_0.png","tag":["Tech"],"readingTime":10},"content":"\u003c!doctype html\u003e\n\u003chtml lang=\"en\"\u003e\n\u003chead\u003e\n\u003cmeta charset=\"utf-8\"\u003e\n\u003cmeta content=\"width=device-width, initial-scale=1\" name=\"viewport\"\u003e\n\u003c/head\u003e\n\u003cbody\u003e\n\u003ch2\u003e사용자에게 경고하기 위해 do-catch 문과 사용자 정의 오류를 던지는 방법 (코드 예제와 함께)\u003c/h2\u003e\n\u003cp\u003e에러 처리는 모든 앱의 기본적인 부분입니다.\u003c/p\u003e\n\u003cp\u003e사용자가 수행하는 모든 조치가 성공하는 것은 아닙니다. 일부는 실패할 것이며, 우리의 앱은 사용자에게 일어난 일을 쉽게 알려주고 문제를 피하기 위해 무엇을 시도할 수 있는지 알려줘야 합니다.\u003c/p\u003e\n\u003cp\u003e간단한 throw 및 do-catch 문을 사용하여 간단한 에러 처리를 어떻게 달성할 수 있는지 보여드릴게요.\u003c/p\u003e\n\u003c!-- ui-station 사각형 --\u003e\n\u003cp\u003e\u003cins class=\"adsbygoogle\" style=\"display:block\" data-ad-client=\"ca-pub-4877378276818686\" data-ad-slot=\"7249294152\" data-ad-format=\"auto\" data-full-width-responsive=\"true\"\u003e\u003c/ins\u003e\u003c/p\u003e\n\u003cscript\u003e\n(adsbygoogle = window.adsbygoogle || []).push({});\n\u003c/script\u003e\n\u003cimg src=\"/assets/img/2024-06-19-에러핸들링인Swif_0.png\"\u003e\n\u003ch2\u003e에러 정의 및 던지기\u003c/h2\u003e\n\u003cp\u003e에러를 표현하려면 Error 프로토콜을 준수하는 타입을 사용해야 합니다. 이는 아무런 제약이 없는 빈 프로토콜이며, 우리가 가장 잘 적합한 타입으로 에러를 정의할 수 있다는 것을 의미합니다.\u003c/p\u003e\n\u003cp\u003e일반적으로 enum이 시작하기에 가장 좋은 방법입니다. 네트워크 에러를 처리하고 싶다고 가정해보면, 아래 enum을 정의할 수 있습니다👇\u003c/p\u003e\n\u003c!-- ui-station 사각형 --\u003e\n\u003cp\u003e\u003cins class=\"adsbygoogle\" style=\"display:block\" data-ad-client=\"ca-pub-4877378276818686\" data-ad-slot=\"7249294152\" data-ad-format=\"auto\" data-full-width-responsive=\"true\"\u003e\u003c/ins\u003e\u003c/p\u003e\n\u003cscript\u003e\n(adsbygoogle = window.adsbygoogle || []).push({});\n\u003c/script\u003e\n\u003cpre\u003e\u003ccode class=\"hljs language-js\"\u003e열거형 \u003cspan class=\"hljs-title class_\"\u003eNetworkError\u003c/span\u003e: 오류 {\n    \u003cspan class=\"hljs-keyword\"\u003ecase\u003c/span\u003e unexpected\n    \u003cspan class=\"hljs-keyword\"\u003ecase\u003c/span\u003e \u003cspan class=\"hljs-title function_\"\u003einvalidURL\u003c/span\u003e(_ \u003cspan class=\"hljs-attr\"\u003eurl\u003c/span\u003e: \u003cspan class=\"hljs-variable constant_\"\u003eURL\u003c/span\u003e)\n    \u003cspan class=\"hljs-keyword\"\u003ecase\u003c/span\u003e \u003cspan class=\"hljs-title function_\"\u003eapiError\u003c/span\u003e(\u003cspan class=\"hljs-attr\"\u003estatusCode\u003c/span\u003e: \u003cspan class=\"hljs-title class_\"\u003eInt\u003c/span\u003e)\n}\n\u003c/code\u003e\u003c/pre\u003e\n\u003cp\u003e백엔드 API로 요청을 수행할 때 오류가 발생하면 NetworkError를 던질 수 있습니다.\u003c/p\u003e\n\u003cpre\u003e\u003ccode class=\"hljs language-js\"\u003e열거형 \u003cspan class=\"hljs-title class_\"\u003eApiMethod\u003c/span\u003e {...}\n구조체 \u003cspan class=\"hljs-title class_\"\u003eApiRequest\u003c/span\u003e {...}\n\nfunc \u003cspan class=\"hljs-title function_\"\u003ecreateRequest\u003c/span\u003e(\u003cspan class=\"hljs-keyword\"\u003efrom\u003c/span\u003e \u003cspan class=\"hljs-attr\"\u003eurl\u003c/span\u003e: \u003cspan class=\"hljs-title class_\"\u003eString\u003c/span\u003e, \u003cspan class=\"hljs-attr\"\u003emethod\u003c/span\u003e: \u003cspan class=\"hljs-title class_\"\u003eApiMethod\u003c/span\u003e) throws -\u003e \u003cspan class=\"hljs-title class_\"\u003eApiRequest\u003c/span\u003e {\n  guard \u003cspan class=\"hljs-keyword\"\u003elet\u003c/span\u003e \u003cspan class=\"hljs-variable constant_\"\u003eURL\u003c/span\u003e = \u003cspan class=\"hljs-title function_\"\u003eURL\u003c/span\u003e(\u003cspan class=\"hljs-attr\"\u003estring\u003c/span\u003e: url) \u003cspan class=\"hljs-keyword\"\u003eelse\u003c/span\u003e {\n    \u003cspan class=\"hljs-keyword\"\u003ethrow\u003c/span\u003e \u003cspan class=\"hljs-title class_\"\u003eNetworkError\u003c/span\u003e.\u003cspan class=\"hljs-title function_\"\u003einvalidURL\u003c/span\u003e(url)\n  }\n  ...\n}\n\u003c/code\u003e\u003c/pre\u003e\n\u003ch2\u003e오류 처리\u003c/h2\u003e\n\u003c!-- ui-station 사각형 --\u003e\n\u003cp\u003e\u003cins class=\"adsbygoogle\" style=\"display:block\" data-ad-client=\"ca-pub-4877378276818686\" data-ad-slot=\"7249294152\" data-ad-format=\"auto\" data-full-width-responsive=\"true\"\u003e\u003c/ins\u003e\u003c/p\u003e\n\u003cscript\u003e\n(adsbygoogle = window.adsbygoogle || []).push({});\n\u003c/script\u003e\n\u003cp\u003e우리는 사용자 정의 오류를 갖고 있고, 이를 throw하고 있습니다. 하지만, 이를 어떻게 잡을 수 있을까요?\u003c/p\u003e\n\u003cp\u003e다른 많은 언어와 마찬가지로, Swift에는 do-catch 문이 있습니다. 만약 do 절 내에서 오류가 throw되면, 실행은 catch 절로 전달되고 오류는 로컬 상수 error에 저장됩니다.\u003c/p\u003e\n\u003cpre\u003e\u003ccode class=\"hljs language-js\"\u003e\u003cspan class=\"hljs-keyword\"\u003edo\u003c/span\u003e {\n  \u003cspan class=\"hljs-keyword\"\u003elet\u003c/span\u003e result = \u003cspan class=\"hljs-keyword\"\u003etry\u003c/span\u003e \u003cspan class=\"hljs-title function_\"\u003ecreateRequest\u003c/span\u003e(\u003cspan class=\"hljs-attr\"\u003efrom\u003c/span\u003e: \u003cspan class=\"hljs-string\"\u003e\"//myapi.com/\"\u003c/span\u003e, \u003cspan class=\"hljs-attr\"\u003emethod\u003c/span\u003e: .\u003cspan class=\"hljs-property\"\u003eGET\u003c/span\u003e)\n} \u003cspan class=\"hljs-keyword\"\u003ecatch\u003c/span\u003e {\n  \u003cspan class=\"hljs-title function_\"\u003eprint\u003c/span\u003e(error)\n  \u003cspan class=\"hljs-comment\"\u003e// 적절한 처리를 수행\u003c/span\u003e\n}\n\u003c/code\u003e\u003c/pre\u003e\n\u003cp\u003e가끔씩 우리는 잡은 오류 형식(또는 오류 값)에 따라 특정 작업을 트리거하길 원합니다. 이를 패턴을 사용하여 수행할 수 있습니다.\u003c/p\u003e\n\u003c!-- ui-station 사각형 --\u003e\n\u003cp\u003e\u003cins class=\"adsbygoogle\" style=\"display:block\" data-ad-client=\"ca-pub-4877378276818686\" data-ad-slot=\"7249294152\" data-ad-format=\"auto\" data-full-width-responsive=\"true\"\u003e\u003c/ins\u003e\u003c/p\u003e\n\u003cscript\u003e\n(adsbygoogle = window.adsbygoogle || []).push({});\n\u003c/script\u003e\n\u003cpre\u003e\u003ccode class=\"hljs language-js\"\u003e\u003cspan class=\"hljs-keyword\"\u003edo\u003c/span\u003e {\n  \u003cspan class=\"hljs-keyword\"\u003elet\u003c/span\u003e result = \u003cspan class=\"hljs-keyword\"\u003etry\u003c/span\u003e \u003cspan class=\"hljs-title function_\"\u003emakeRequest\u003c/span\u003e(\u003cspan class=\"hljs-attr\"\u003eto\u003c/span\u003e: \u003cspan class=\"hljs-string\"\u003e\"//myapi.com/\"\u003c/span\u003e, \u003cspan class=\"hljs-attr\"\u003emethod\u003c/span\u003e: .\u003cspan class=\"hljs-property\"\u003eGET\u003c/span\u003e)\n} \u003cspan class=\"hljs-keyword\"\u003ecatch\u003c/span\u003e \u003cspan class=\"hljs-title class_\"\u003eNetworkError\u003c/span\u003e.\u003cspan class=\"hljs-title function_\"\u003einvalidUrl\u003c/span\u003e(\u003cspan class=\"hljs-params\"\u003e\u003cspan class=\"hljs-keyword\"\u003elet\u003c/span\u003e url\u003c/span\u003e) {\n  \u003cspan class=\"hljs-comment\"\u003e// ...\u003c/span\u003e\n} \u003cspan class=\"hljs-keyword\"\u003ecatch\u003c/span\u003e {\n  \u003cspan class=\"hljs-title function_\"\u003eprint\u003c/span\u003e(error)\n  \u003cspan class=\"hljs-comment\"\u003e//\u003c/span\u003e\n}\n\u003c/code\u003e\u003c/pre\u003e\n\u003ch2\u003e에러 전파하기\u003c/h2\u003e\n\u003cp\u003e던져진 함수는 함수 내에서 발생한 에러를 호출자의 범위로 전파할 수 있습니다.\u003c/p\u003e\n\u003cpre\u003e\u003ccode class=\"hljs language-js\"\u003estruct \u003cspan class=\"hljs-title class_\"\u003eUser\u003c/span\u003e: \u003cspan class=\"hljs-title class_\"\u003eDecodable\u003c/span\u003e {\n    \u003cspan class=\"hljs-keyword\"\u003elet\u003c/span\u003e \u003cspan class=\"hljs-attr\"\u003ename\u003c/span\u003e: \u003cspan class=\"hljs-title class_\"\u003eString\u003c/span\u003e\n    \u003cspan class=\"hljs-keyword\"\u003elet\u003c/span\u003e \u003cspan class=\"hljs-attr\"\u003eage\u003c/span\u003e: \u003cspan class=\"hljs-title class_\"\u003eInt\u003c/span\u003e\n}\n\nfunc decodeJSON\u0026#x3C;\u003cspan class=\"hljs-attr\"\u003eT\u003c/span\u003e: \u003cspan class=\"hljs-title class_\"\u003eDecodable\u003c/span\u003e\u003e(_ \u003cspan class=\"hljs-attr\"\u003ejsonString\u003c/span\u003e: \u003cspan class=\"hljs-title class_\"\u003eString\u003c/span\u003e) throws -\u003e T {\n    \u003cspan class=\"hljs-keyword\"\u003elet\u003c/span\u003e jsonData = \u003cspan class=\"hljs-title class_\"\u003eData\u003c/span\u003e(jsonString.\u003cspan class=\"hljs-property\"\u003eutf8\u003c/span\u003e)\n    \u003cspan class=\"hljs-keyword\"\u003elet\u003c/span\u003e decodedObject = \u003cspan class=\"hljs-keyword\"\u003etry\u003c/span\u003e \u003cspan class=\"hljs-title class_\"\u003eJSON\u003c/span\u003eDecoder().\u003cspan class=\"hljs-title function_\"\u003edecode\u003c/span\u003e(T.\u003cspan class=\"hljs-property\"\u003eself\u003c/span\u003e, \u003cspan class=\"hljs-attr\"\u003efrom\u003c/span\u003e: jsonData)\n    \u003cspan class=\"hljs-keyword\"\u003ereturn\u003c/span\u003e decodedObject\n}\n\nfunc \u003cspan class=\"hljs-title function_\"\u003egetUserFromJSON\u003c/span\u003e(_ \u003cspan class=\"hljs-attr\"\u003ejsonString\u003c/span\u003e: \u003cspan class=\"hljs-title class_\"\u003eString\u003c/span\u003e) -\u003e \u003cspan class=\"hljs-title class_\"\u003eUser\u003c/span\u003e? {\n    \u003cspan class=\"hljs-keyword\"\u003edo\u003c/span\u003e {\n        \u003cspan class=\"hljs-keyword\"\u003elet\u003c/span\u003e \u003cspan class=\"hljs-attr\"\u003euser\u003c/span\u003e: \u003cspan class=\"hljs-title class_\"\u003eUser\u003c/span\u003e = \u003cspan class=\"hljs-keyword\"\u003etry\u003c/span\u003e \u003cspan class=\"hljs-title function_\"\u003edecodeJSON\u003c/span\u003e(jsonString)\n        \u003cspan class=\"hljs-keyword\"\u003ereturn\u003c/span\u003e user\n    } \u003cspan class=\"hljs-keyword\"\u003ecatch\u003c/span\u003e {\n        \u003cspan class=\"hljs-title function_\"\u003eprint\u003c/span\u003e(error)\n        \u003cspan class=\"hljs-keyword\"\u003ereturn\u003c/span\u003e nil\n    }\n}\n\u003c/code\u003e\u003c/pre\u003e\n\u003c!-- ui-station 사각형 --\u003e\n\u003cp\u003e\u003cins class=\"adsbygoogle\" style=\"display:block\" data-ad-client=\"ca-pub-4877378276818686\" data-ad-slot=\"7249294152\" data-ad-format=\"auto\" data-full-width-responsive=\"true\"\u003e\u003c/ins\u003e\u003c/p\u003e\n\u003cscript\u003e\n(adsbygoogle = window.adsbygoogle || []).push({});\n\u003c/script\u003e\n\u003ch2\u003etry? \u0026#x26; try!\u003c/h2\u003e\n\u003cp\u003e가끔은 던지는 함수로 발생한 오류에 대해 신경 쓰지 않을 수 있습니다. 이런 상황에서는 do-catch 문을 사용하지 않고 try? 지시어를 사용할 수 있습니다.\u003c/p\u003e\n\u003cp\u003etry?를 사용하면 호출하는 함수가 옵셔널이 됩니다. 즉, 반환 값을 가진 함수를 호출하는 경우 값은 nil이 됩니다.\u003c/p\u003e\n\u003cpre\u003e\u003ccode class=\"hljs language-js\"\u003e\u003cspan class=\"hljs-keyword\"\u003elet\u003c/span\u003e request = \u003cspan class=\"hljs-keyword\"\u003etry\u003c/span\u003e? \u003cspan class=\"hljs-title function_\"\u003ecreateRequest\u003c/span\u003e(\u003cspan class=\"hljs-attr\"\u003efrom\u003c/span\u003e: \u003cspan class=\"hljs-string\"\u003e\"malformedurl\"\u003c/span\u003e, \u003cspan class=\"hljs-attr\"\u003emethod\u003c/span\u003e: .\u003cspan class=\"hljs-property\"\u003eGET\u003c/span\u003e)\n\u003cspan class=\"hljs-comment\"\u003e// request = nil\u003c/span\u003e\n\u003c/code\u003e\u003c/pre\u003e\n\u003c!-- ui-station 사각형 --\u003e\n\u003cp\u003e\u003cins class=\"adsbygoogle\" style=\"display:block\" data-ad-client=\"ca-pub-4877378276818686\" data-ad-slot=\"7249294152\" data-ad-format=\"auto\" data-full-width-responsive=\"true\"\u003e\u003c/ins\u003e\u003c/p\u003e\n\u003cscript\u003e\n(adsbygoogle = window.adsbygoogle || []).push({});\n\u003c/script\u003e\n\u003cp\u003e한편, 때때로 특정한 던지기 함수가 런타임에서 실패하지 않을 것을 확신할 수 있습니다. 이런 경우에는 try!를 사용할 수 있습니다. 그리고 try?와 마찬가지로 do-catch 클로저를 사용할 필요가 없습니다.\u003c/p\u003e\n\u003cpre\u003e\u003ccode class=\"hljs language-js\"\u003e\u003cspan class=\"hljs-keyword\"\u003elet\u003c/span\u003e request = \u003cspan class=\"hljs-keyword\"\u003etry\u003c/span\u003e! \u003cspan class=\"hljs-title function_\"\u003ecreateRequest\u003c/span\u003e(\u003cspan class=\"hljs-attr\"\u003efrom\u003c/span\u003e: \u003cspan class=\"hljs-string\"\u003e\"https://myapi.com/\"\u003c/span\u003e, \u003cspan class=\"hljs-attr\"\u003emethod\u003c/span\u003e: .\u003cspan class=\"hljs-property\"\u003eGET\u003c/span\u003e)\n\u003c/code\u003e\u003c/pre\u003e\n\u003ch2\u003e비동기 코드\u003c/h2\u003e\n\u003cp\u003e에러를 던지는 것은 비동기 코드와 잘 작동합니다. 함수 선언에서 throws 키워드 앞에 async 키워드를 넣어주어야 합니다.\u003c/p\u003e\n\u003c!-- ui-station 사각형 --\u003e\n\u003cp\u003e\u003cins class=\"adsbygoogle\" style=\"display:block\" data-ad-client=\"ca-pub-4877378276818686\" data-ad-slot=\"7249294152\" data-ad-format=\"auto\" data-full-width-responsive=\"true\"\u003e\u003c/ins\u003e\u003c/p\u003e\n\u003cscript\u003e\n(adsbygoogle = window.adsbygoogle || []).push({});\n\u003c/script\u003e\n\u003cp\u003e비동기 함수를 호출할 때는 try await (키워드 순서 스와이프)를 사용합니다.\u003c/p\u003e\n\u003cpre\u003e\u003ccode class=\"hljs language-js\"\u003efunc \u003cspan class=\"hljs-title function_\"\u003eperformRequest\u003c/span\u003e(\u003cspan class=\"hljs-keyword\"\u003ewith\u003c/span\u003e \u003cspan class=\"hljs-attr\"\u003erequest\u003c/span\u003e: \u003cspan class=\"hljs-title class_\"\u003eApiRequest\u003c/span\u003e) \u003cspan class=\"hljs-keyword\"\u003easync\u003c/span\u003e throws -\u003e \u003cspan class=\"hljs-title class_\"\u003eApiResult\u003c/span\u003e {...}\n\n\u003cspan class=\"hljs-comment\"\u003e// ----------------\u003c/span\u003e\n\n\u003cspan class=\"hljs-keyword\"\u003edo\u003c/span\u003e {\n  \u003cspan class=\"hljs-keyword\"\u003elet\u003c/span\u003e request = \u003cspan class=\"hljs-keyword\"\u003etry\u003c/span\u003e \u003cspan class=\"hljs-title function_\"\u003emakeRequest\u003c/span\u003e(\u003cspan class=\"hljs-attr\"\u003eto\u003c/span\u003e: \u003cspan class=\"hljs-string\"\u003e\"https://myapi.com/users\"\u003c/span\u003e, \u003cspan class=\"hljs-attr\"\u003emethod\u003c/span\u003e: .\u003cspan class=\"hljs-property\"\u003eGET\u003c/span\u003e)\n  \u003cspan class=\"hljs-keyword\"\u003elet\u003c/span\u003e apiRestul = \u003cspan class=\"hljs-keyword\"\u003etry\u003c/span\u003e \u003cspan class=\"hljs-keyword\"\u003eawait\u003c/span\u003e \u003cspan class=\"hljs-title function_\"\u003eperformRequest\u003c/span\u003e(\u003cspan class=\"hljs-attr\"\u003ewith\u003c/span\u003e: request)\n} \u003cspan class=\"hljs-keyword\"\u003ecatch\u003c/span\u003e \u003cspan class=\"hljs-title class_\"\u003eNetworkError\u003c/span\u003e.\u003cspan class=\"hljs-title function_\"\u003einvalidUrl\u003c/span\u003e(\u003cspan class=\"hljs-params\"\u003e\u003cspan class=\"hljs-keyword\"\u003elet\u003c/span\u003e url\u003c/span\u003e) {\n  \u003cspan class=\"hljs-comment\"\u003e// ...\u003c/span\u003e\n} \u003cspan class=\"hljs-keyword\"\u003ecatch\u003c/span\u003e {\n  \u003cspan class=\"hljs-title function_\"\u003eprint\u003c/span\u003e(error)\n  \u003cspan class=\"hljs-comment\"\u003e//\u003c/span\u003e\n}\n\u003c/code\u003e\u003c/pre\u003e\n\u003ch1\u003e사용자에게 알림\u003c/h1\u003e\n\u003cp\u003e에러 구조를 구축했으니, 이제 사용자들에게 무엇이 잘못되었는지 알려줄 필요가 있습니다.\u003c/p\u003e\n\u003c!-- ui-station 사각형 --\u003e\n\u003cp\u003e\u003cins class=\"adsbygoogle\" style=\"display:block\" data-ad-client=\"ca-pub-4877378276818686\" data-ad-slot=\"7249294152\" data-ad-format=\"auto\" data-full-width-responsive=\"true\"\u003e\u003c/ins\u003e\u003c/p\u003e\n\u003cscript\u003e\n(adsbygoogle = window.adsbygoogle || []).push({});\n\u003c/script\u003e\n\u003cimg src=\"https://miro.medium.com/v2/resize:fit:592/1*cfrDQghry5EPpPhMSFFDaA.gif\"\u003e\n\u003cp\u003e각각의 오류에 대한 설명 은 오류가 처음에 발생한 이유입니다.\n사용자가 사용자를 피할 수 있는 가능한 조치입니다.\u003c/p\u003e\n\u003cp\u003e우리는 Error enum에 해당 정보를 변수로 추가할 수 있습니다. 그러나 이미 모든 이 정보를 그룹화하기 위한 프로토콜이 있습니다: LocalizedError. 우리의 사용 사례에서 우리는 failureReason 및 recoverySuggestion 변수 만 구현하면 됩니다.\u003c/p\u003e\n\u003c!-- ui-station 사각형 --\u003e\n\u003cp\u003e\u003cins class=\"adsbygoogle\" style=\"display:block\" data-ad-client=\"ca-pub-4877378276818686\" data-ad-slot=\"7249294152\" data-ad-format=\"auto\" data-full-width-responsive=\"true\"\u003e\u003c/ins\u003e\u003c/p\u003e\n\u003cscript\u003e\n(adsbygoogle = window.adsbygoogle || []).push({});\n\u003c/script\u003e\n\u003cpre\u003e\u003ccode class=\"hljs language-swift\"\u003e\u003cspan class=\"hljs-keyword\"\u003eenum\u003c/span\u003e \u003cspan class=\"hljs-title class_\"\u003eOrderError\u003c/span\u003e: \u003cspan class=\"hljs-title class_\"\u003eLocalizedError\u003c/span\u003e {\n  \u003cspan class=\"hljs-keyword\"\u003ecase\u003c/span\u003e unexpected\n  \u003cspan class=\"hljs-keyword\"\u003ecase\u003c/span\u003e outOfCoffee(coffee: \u003cspan class=\"hljs-type\"\u003eCoffee\u003c/span\u003e)\n  \u003cspan class=\"hljs-keyword\"\u003ecase\u003c/span\u003e minimumNotMet(currentPrice: \u003cspan class=\"hljs-type\"\u003eFloat\u003c/span\u003e)\n\n  \u003cspan class=\"hljs-keyword\"\u003evar\u003c/span\u003e failureReason: \u003cspan class=\"hljs-type\"\u003eString\u003c/span\u003e? {\n    \u003cspan class=\"hljs-keyword\"\u003eswitch\u003c/span\u003e \u003cspan class=\"hljs-keyword\"\u003eself\u003c/span\u003e {\n    \u003cspan class=\"hljs-keyword\"\u003ecase\u003c/span\u003e .unexpected:\n      \u003cspan class=\"hljs-keyword\"\u003ereturn\u003c/span\u003e \u003cspan class=\"hljs-string\"\u003e\"주문 처리 중 문제가 발생했습니다.\"\u003c/span\u003e\n    \u003cspan class=\"hljs-keyword\"\u003ecase\u003c/span\u003e .outOfCoffee(\u003cspan class=\"hljs-keyword\"\u003elet\u003c/span\u003e coffee):\n      \u003cspan class=\"hljs-keyword\"\u003ereturn\u003c/span\u003e \u003cspan class=\"hljs-string\"\u003e\"\u003cspan class=\"hljs-subst\"\u003e\\(coffee.name)\u003c/span\u003e가 다 떨어졌어요.\"\u003c/span\u003e\n    \u003cspan class=\"hljs-keyword\"\u003ecase\u003c/span\u003e .minimumNotMet:\n      \u003cspan class=\"hljs-keyword\"\u003ereturn\u003c/span\u003e \u003cspan class=\"hljs-string\"\u003e\"앱에서 주문하려면 최소 $10 이상 사용해야 합니다.\"\u003c/span\u003e\n    }\n  }\n\n  \u003cspan class=\"hljs-keyword\"\u003evar\u003c/span\u003e recoverySuggestion: \u003cspan class=\"hljs-type\"\u003eString\u003c/span\u003e? {\n    \u003cspan class=\"hljs-keyword\"\u003eswitch\u003c/span\u003e \u003cspan class=\"hljs-keyword\"\u003eself\u003c/span\u003e {\n    \u003cspan class=\"hljs-keyword\"\u003ecase\u003c/span\u003e .unexpected:\n      \u003cspan class=\"hljs-keyword\"\u003ereturn\u003c/span\u003e \u003cspan class=\"hljs-string\"\u003e\"몇 분 후에 다시 시도해주세요.\"\u003c/span\u003e\n    \u003cspan class=\"hljs-keyword\"\u003ecase\u003c/span\u003e .outOfCoffee:\n      \u003cspan class=\"hljs-keyword\"\u003ereturn\u003c/span\u003e \u003cspan class=\"hljs-string\"\u003e\"다른 종류의 커피로 변경해보세요.\"\u003c/span\u003e\n    \u003cspan class=\"hljs-keyword\"\u003ecase\u003c/span\u003e .minimumNotMet(\u003cspan class=\"hljs-keyword\"\u003elet\u003c/span\u003e currentPrice):\n      \u003cspan class=\"hljs-keyword\"\u003ereturn\u003c/span\u003e \u003cspan class=\"hljs-string\"\u003e\"주문에 $\u003cspan class=\"hljs-subst\"\u003e\\(\u003cspan class=\"hljs-number\"\u003e10\u003c/span\u003e\u003cspan class=\"hljs-operator\"\u003e-\u003c/span\u003ecurrentPrice)\u003c/span\u003e 더 추가해주세요.\"\u003c/span\u003e\n    }\n  }\n}\n\u003c/code\u003e\u003c/pre\u003e\n\u003cp\u003e여기에서 작업이 완료됩니다. 그러나 앱에 통합된 로깅 솔루션이 있다면, 오류 구조를 확장하여 추가 정보를 추가할 수 있습니다.\u003c/p\u003e\n\u003cpre\u003e\u003ccode class=\"hljs language-swift\"\u003e\u003cspan class=\"hljs-keyword\"\u003eimport\u003c/span\u003e Foundation\n\u003cspan class=\"hljs-keyword\"\u003eimport\u003c/span\u003e HKLogger\n\n\u003cspan class=\"hljs-keyword\"\u003estruct\u003c/span\u003e \u003cspan class=\"hljs-title class_\"\u003eAppError\u003c/span\u003e: \u003cspan class=\"hljs-title class_\"\u003eError\u003c/span\u003e {\n  \u003cspan class=\"hljs-keyword\"\u003elet\u003c/span\u003e type: \u003cspan class=\"hljs-type\"\u003eLocalizedError\u003c/span\u003e\n  \u003cspan class=\"hljs-keyword\"\u003evar\u003c/span\u003e userMessage: \u003cspan class=\"hljs-type\"\u003eString\u003c/span\u003e {\n    \u003cspan class=\"hljs-keyword\"\u003ereturn\u003c/span\u003e \u003cspan class=\"hljs-string\"\u003e\"\u003cspan class=\"hljs-subst\"\u003e\\(type.failureReason \u003cspan class=\"hljs-operator\"\u003e??\u003c/span\u003e ErrorConstants.defaultError)\u003c/span\u003e \u003cspan class=\"hljs-subst\"\u003e\\n\u003c/span\u003e\u003cspan class=\"hljs-subst\"\u003e\\n\u003c/span\u003e \u003cspan class=\"hljs-subst\"\u003e\\(type.recoverySuggestion \u003cspan class=\"hljs-operator\"\u003e??\u003c/span\u003e ErrorConstants.defaultAction)\u003c/span\u003e\"\u003c/span\u003e\n  }\n  \u003cspan class=\"hljs-comment\"\u003e// 필요한 만큼 많은 속성을 추가\u003c/span\u003e\n\n  \u003cspan class=\"hljs-keyword\"\u003einit\u003c/span\u003e(\u003cspan class=\"hljs-params\"\u003etype\u003c/span\u003e: \u003cspan class=\"hljs-type\"\u003eLocalizedError\u003c/span\u003e, \u003cspan class=\"hljs-params\"\u003edebugInfo\u003c/span\u003e: \u003cspan class=\"hljs-type\"\u003eString\u003c/span\u003e? \u003cspan class=\"hljs-operator\"\u003e=\u003c/span\u003e \u003cspan class=\"hljs-literal\"\u003enil\u003c/span\u003e) {\n      \u003cspan class=\"hljs-keyword\"\u003eself\u003c/span\u003e.type \u003cspan class=\"hljs-operator\"\u003e=\u003c/span\u003e type\n      \u003cspan class=\"hljs-keyword\"\u003eguard\u003c/span\u003e \u003cspan class=\"hljs-keyword\"\u003elet\u003c/span\u003e debugInfo \u003cspan class=\"hljs-keyword\"\u003eelse\u003c/span\u003e { \u003cspan class=\"hljs-keyword\"\u003ereturn\u003c/span\u003e }\n      \u003cspan class=\"hljs-type\"\u003eHKLogger\u003c/span\u003e.shared.error(message: debugInfo)\n  }\n}\n\u003c/code\u003e\u003c/pre\u003e\n\u003cp\u003e이렇게 함으로써 로깅 오류 로직을 한 곳에 중앙 집중화할 수 있습니다.\u003c/p\u003e\n\u003c!-- ui-station 사각형 --\u003e\n\u003cp\u003e\u003cins class=\"adsbygoogle\" style=\"display:block\" data-ad-client=\"ca-pub-4877378276818686\" data-ad-slot=\"7249294152\" data-ad-format=\"auto\" data-full-width-responsive=\"true\"\u003e\u003c/ins\u003e\u003c/p\u003e\n\u003cscript\u003e\n(adsbygoogle = window.adsbygoogle || []).push({});\n\u003c/script\u003e\n\u003cp\u003e마침내 사용자 정의 AppError를 사용하여 경고를 표시하는 간단한 View 확장을 만들 수 있습니다.\u003c/p\u003e\n\u003cpre\u003e\u003ccode class=\"hljs language-js\"\u003eextension \u003cspan class=\"hljs-title class_\"\u003eView\u003c/span\u003e {\n    func \u003cspan class=\"hljs-title function_\"\u003ealert\u003c/span\u003e(\u003cspan class=\"hljs-attr\"\u003eisPresented\u003c/span\u003e: \u003cspan class=\"hljs-title class_\"\u003eBinding\u003c/span\u003e\u0026#x3C;\u003cspan class=\"hljs-title class_\"\u003eBool\u003c/span\u003e\u003e, withError \u003cspan class=\"hljs-attr\"\u003eerror\u003c/span\u003e: \u003cspan class=\"hljs-title class_\"\u003eAppError\u003c/span\u003e?) -\u003e some \u003cspan class=\"hljs-title class_\"\u003eView\u003c/span\u003e {\n        \u003cspan class=\"hljs-keyword\"\u003ereturn\u003c/span\u003e \u003cspan class=\"hljs-title function_\"\u003ealert\u003c/span\u003e(\n            \u003cspan class=\"hljs-string\"\u003e\"Ups! :(\"\u003c/span\u003e,\n            \u003cspan class=\"hljs-attr\"\u003eisPresented\u003c/span\u003e: isPresented,\n            \u003cspan class=\"hljs-attr\"\u003eactions\u003c/span\u003e: {\n                \u003cspan class=\"hljs-title class_\"\u003eButton\u003c/span\u003e(\u003cspan class=\"hljs-string\"\u003e\"Ok\"\u003c/span\u003e) {}\n            }, \u003cspan class=\"hljs-attr\"\u003emessage\u003c/span\u003e: {\n                \u003cspan class=\"hljs-title class_\"\u003eText\u003c/span\u003e(error?.\u003cspan class=\"hljs-property\"\u003euserMessage\u003c/span\u003e ?? \u003cspan class=\"hljs-string\"\u003e\"\"\u003c/span\u003e)\n            }\n        )\n    }\n}\n\n\u003cspan class=\"hljs-comment\"\u003e// ---------------------------------------------------\u003c/span\u003e\n\nstruct \u003cspan class=\"hljs-title class_\"\u003eOrderConfirmationView\u003c/span\u003e: \u003cspan class=\"hljs-title class_\"\u003eView\u003c/span\u003e {\n  @\u003cspan class=\"hljs-title class_\"\u003eStateObject\u003c/span\u003e private \u003cspan class=\"hljs-keyword\"\u003evar\u003c/span\u003e viewModel = \u003cspan class=\"hljs-title class_\"\u003eConfirmationViewModel\u003c/span\u003e()\n\n  \u003cspan class=\"hljs-keyword\"\u003evar\u003c/span\u003e \u003cspan class=\"hljs-attr\"\u003ebody\u003c/span\u003e: some \u003cspan class=\"hljs-title class_\"\u003eView\u003c/span\u003e {\n    \u003cspan class=\"hljs-title class_\"\u003eVStack\u003c/span\u003e(\u003cspan class=\"hljs-attr\"\u003ealignment\u003c/span\u003e: .\u003cspan class=\"hljs-property\"\u003eleading\u003c/span\u003e) {\n      \u003cspan class=\"hljs-comment\"\u003e// ...\u003c/span\u003e\n      \u003cspan class=\"hljs-title class_\"\u003eButton\u003c/span\u003e(\u003cspan class=\"hljs-string\"\u003e\"주문하기\"\u003c/span\u003e) {\n        viewModel.\u003cspan class=\"hljs-title function_\"\u003etryToPlaceOrder\u003c/span\u003e(order)\n      }\n    }\n    .\u003cspan class=\"hljs-title function_\"\u003ealert\u003c/span\u003e(\u003cspan class=\"hljs-attr\"\u003eisPresented\u003c/span\u003e: $viewModel.\u003cspan class=\"hljs-property\"\u003eshowError\u003c/span\u003e, \u003cspan class=\"hljs-attr\"\u003ewithError\u003c/span\u003e: viewModel.\u003cspan class=\"hljs-property\"\u003eappError\u003c/span\u003e)\n  }\n}\n\n\u003cspan class=\"hljs-comment\"\u003e// ---------------------------------------------------\u003c/span\u003e\n\n@\u003cspan class=\"hljs-title class_\"\u003eObservable\u003c/span\u003e\nfinal \u003cspan class=\"hljs-keyword\"\u003eclass\u003c/span\u003e \u003cspan class=\"hljs-title class_\"\u003eConfirmationViewModel\u003c/span\u003e: \u003cspan class=\"hljs-title class_\"\u003eObservableObject\u003c/span\u003e {\n    \u003cspan class=\"hljs-keyword\"\u003evar\u003c/span\u003e \u003cspan class=\"hljs-attr\"\u003eappError\u003c/span\u003e: \u003cspan class=\"hljs-title class_\"\u003eAppError\u003c/span\u003e?\n    \u003cspan class=\"hljs-keyword\"\u003evar\u003c/span\u003e showError = \u003cspan class=\"hljs-literal\"\u003efalse\u003c/span\u003e\n\n    func \u003cspan class=\"hljs-title function_\"\u003etryToPlaceOrder\u003c/span\u003e(\u003cspan class=\"hljs-params\"\u003e_ order: Order\u003c/span\u003e) {\n        \u003cspan class=\"hljs-keyword\"\u003edo\u003c/span\u003e {\n            \u003cspan class=\"hljs-keyword\"\u003etry\u003c/span\u003e \u003cspan class=\"hljs-title class_\"\u003eOrdersManager\u003c/span\u003e.\u003cspan class=\"hljs-property\"\u003eshared\u003c/span\u003e.\u003cspan class=\"hljs-title function_\"\u003eadd\u003c/span\u003e(order)\n        } \u003cspan class=\"hljs-keyword\"\u003ecatch\u003c/span\u003e \u003cspan class=\"hljs-keyword\"\u003elet\u003c/span\u003e error \u003cspan class=\"hljs-keyword\"\u003eas\u003c/span\u003e \u003cspan class=\"hljs-title class_\"\u003eAppError\u003c/span\u003e {\n            showError = \u003cspan class=\"hljs-literal\"\u003etrue\u003c/span\u003e\n            appError = error\n        } \u003cspan class=\"hljs-keyword\"\u003ecatch\u003c/span\u003e {\n            appError = \u003cspan class=\"hljs-title class_\"\u003eAppError\u003c/span\u003e(\u003cspan class=\"hljs-attr\"\u003etype\u003c/span\u003e: \u003cspan class=\"hljs-title class_\"\u003eOrderError\u003c/span\u003e.\u003cspan class=\"hljs-property\"\u003eunexpected\u003c/span\u003e)\n        }\n    }\n}\n\n\u003cspan class=\"hljs-comment\"\u003e// ---------------------------------------------------\u003c/span\u003e\n\nfinal \u003cspan class=\"hljs-keyword\"\u003eclass\u003c/span\u003e \u003cspan class=\"hljs-title class_\"\u003eOrdersManager\u003c/span\u003e {\n  \u003cspan class=\"hljs-comment\"\u003e// ...\u003c/span\u003e\n  func \u003cspan class=\"hljs-title function_\"\u003eadd\u003c/span\u003e(_ \u003cspan class=\"hljs-attr\"\u003eorder\u003c/span\u003e: \u003cspan class=\"hljs-title class_\"\u003eOrder\u003c/span\u003e) throws {\n    \u003cspan class=\"hljs-keyword\"\u003eif\u003c/span\u003e order.\u003cspan class=\"hljs-property\"\u003eprice\u003c/span\u003e \u003e= \u003cspan class=\"hljs-number\"\u003e10\u003c/span\u003e {\n      orders.\u003cspan class=\"hljs-title function_\"\u003eappend\u003c/span\u003e(order)\n    } \u003cspan class=\"hljs-keyword\"\u003eelse\u003c/span\u003e {\n      \u003cspan class=\"hljs-keyword\"\u003ethrow\u003c/span\u003e \u003cspan class=\"hljs-title class_\"\u003eAppError\u003c/span\u003e(\u003cspan class=\"hljs-attr\"\u003etype\u003c/span\u003e: \u003cspan class=\"hljs-title class_\"\u003eOrderError\u003c/span\u003e.\u003cspan class=\"hljs-title function_\"\u003eminimumNotMet\u003c/span\u003e(\u003cspan class=\"hljs-attr\"\u003ecurrentPrice\u003c/span\u003e: order.\u003cspan class=\"hljs-property\"\u003eprice\u003c/span\u003e))\n    }\n  }\n}\n\u003c/code\u003e\u003c/pre\u003e\n\u003cp\u003e질문이 있으신가요? 언제든지 메시지를 남겨주세요! 🙂\u003c/p\u003e\n\u003cul\u003e\n\u003cli\u003e🤓 iOS 개발 팁과 통찰력에 관한 콘텐츠를 X에서 함께 확인해보세요.\u003c/li\u003e\n\u003cli\u003e🚀 모든 예제 프로젝트를 공유하는 내 GitHub를 확인해보세요 (이것도 포함됩니다 😉)\u003c/li\u003e\n\u003c/ul\u003e\n\u003c/body\u003e\n\u003c/html\u003e\n"},"__N_SSG":true},"page":"/post/[slug]","query":{"slug":"2024-06-19-ErrorhandlinginSwift"},"buildId":"YUMR4jSyk_WlOHHc7UfOk","isFallback":false,"gsp":true,"scriptLoader":[{"async":true,"src":"https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js?client=ca-pub-4877378276818686","strategy":"lazyOnload","crossOrigin":"anonymous"}]}</script></body></html>